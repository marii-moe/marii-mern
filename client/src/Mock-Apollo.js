import { buildClientSchema } from 'graphql';
import * as introspectionResult from './server-schema.json';
import { addMockFunctionsToSchema } from 'graphql-tools'
import ApolloClient from 'apollo-client'
import { ApolloLink, Observable } from "apollo-link";
import { InMemoryCache } from "apollo-cache-inmemory";
import { shallow } from 'enzyme'
import { ApolloProvider } from 'react-apollo'
import React from 'react'

const schema = buildClientSchema(introspectionResult);

addMockFunctionsToSchema({schema});

const link = new ApolloLink(operation => {
  return new Observable(observer => {
    const { query, operationName, variables } = operation;
    delay(100)
      .then(() =>
        graphql(schema, print(query), null, null, variables, operationName)
      )
      .then(result => {
        observer.next(result);
        observer.complete();
      })
      .catch(observer.error.bind(observer));
  });
});

function delay(ms) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}

export const mockClient = new ApolloClient({
  link,
  cache: new InMemoryCache()
})

export const apolloRender = (element) => {
  const wrapper = shallow(<ApolloProvider client={mockClient}>{element}</ApolloProvider>                                                                                                     
                          , { context: { client: mockClient }})                                                                                                                                  
  const renderedElement = wrapper.dive().dive().dive()
  renderedElement.root=wrapper
  return renderedElement
}
