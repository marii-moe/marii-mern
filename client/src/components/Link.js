import React, { Component } from 'react'
import { GC_USER_ID } from '../constants'
import { timeDifferenceForDate } from '../utils'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

class Link extends Component {

  render() {
    const userId = localStorage.getItem(GC_USER_ID)
    return (
      <div className='flex mt2 items-start'>
	<div className="flex items-center">
	  <span className='gray'>{this.props.index + 1}.</span>
	  {userId && <div data-button='vote' className='ml1 gray fl1' onClick={() => this._voteForLink()}>â–²</div>}
	</div>
	<div className='ml1'>
	  <div>{this.props.link.description} ({this.props.link.url})</div>
	  <div className='f6 lh-copy gray'>
	    {this.props.link.votes.length} votes | by { this.props.link.postedBy.name } {timeDifferenceForDate(this.props.link.createdAt)}
	  </div>
	</div>
      </div>
    )
  }

  _voteForLink = async () => {
    const userId = localStorage.getItem(GC_USER_ID)
    const voterIds = this.props.link.votes.map(vote => vote.user.id)
    const linkId = this.props.link.id
    if(voterIds.includes(userId)) {
      console.log('User already voted for this link')
      return
    }
    await this.props.createVoteMutation({
      variables: {
	linkId
      },
      update: this.props.updateStoreAfterVote
    })
  }  
}

const CREATE_VOTE_MUTATION = gql`
  mutation CreateVoteMutation($linkId: ID!) {
    createVote(linkId: $linkId) {
      id
      link {
        id
        votes {
          id
          user {
            id
          }
        }
      }
      user {
        id
      }
    }
  }
`

export { Link as _Link }

export default graphql(CREATE_VOTE_MUTATION, {
  name: 'createVoteMutation'
})(Link)
