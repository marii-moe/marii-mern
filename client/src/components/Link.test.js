import React from 'react'
import Link from './Link'
import { _Link } from './Link'
import { apolloRender } from '../Mock-Apollo'

const testLink = {
  description: 'test description',
  url: 'http://test.com',
  votes: [],
  postedBy: {
    name: 'test-user'
  }
}

it('should render', async () => {
  expect(apolloRender(<Link link={testLink} />)).toMatchSnapshot()
})

it('should allow users to vote', async () => {
  const link = new _Link()
  const createVoteMutation = jest.fn()
  link.props = {
    link: testLink,
    createVoteMutation: createVoteMutation
  }
  
  await link._voteForLink()
  expect(createVoteMutation.mock.calls.length).toEqual(1)
})

it('should not allow users to vote if already voted', async () => {
  const link = new _Link()
  const createVoteMutation = jest.fn()
  const votedTestLink = Object.assign(testLink, { votes: [{user: {id: 1}}]})
  link.props = {
    link: votedTestLink,
    createVoteMutation: createVoteMutation
  }
  global.localStorage.getItem = () => { return 1 }
  console.log = () => {} //suppress logs in test output
  await link._voteForLink()
  expect(createVoteMutation.mock.calls.length).toEqual(0)
})

it('should render', async () => {
  global.localStorage.getItem = () => { return 1 }
  const wrapper = apolloRender(<Link link={testLink} />)
  const createVote = jest.fn()
  wrapper.instance()._voteForLink = createVote
  wrapper.find('div[data-button="vote"]').simulate('click');
  expect(createVote.mock.calls.length).toEqual(1)
})
