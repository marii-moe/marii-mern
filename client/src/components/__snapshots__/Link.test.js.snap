// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="flex mt2 items-start"
>
  <div
    className="flex items-center"
  >
    <span
      className="gray"
    >
      NaN
      .
    </span>
  </div>
  <div
    className="ml1"
  >
    <div>
      test description
       (
      http://test.com
      )
    </div>
    <div
      className="f6 lh-copy gray"
    >
      0
       votes | by 
      test-user
       
      NaN years ago
    </div>
  </div>
</div>
`;
